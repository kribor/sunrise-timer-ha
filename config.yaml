# Location & max sun angle for artificial light
latitude: 60.025369
longitude: 17.669363

fake_noon_hour: 11 # Treat as mid-day
hours_of_full_light: 10
minutes_of_sunrise: 120 # interval for dimming between 0 and 100% (both sunrise and sundown)

dimmer_step: 2
dimmer_controller: telldus
dimmer_max: 50 # Start dimming downwards from here. Although 100 is 100.

telldus:
  dimmer_id: 9611251
