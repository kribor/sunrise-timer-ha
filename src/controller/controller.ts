import {SunriseTimerConfig} from "../svc/config.service";

export interface Controller {
    dimLight(dimLevel: number, cfg: SunriseTimerConfig): void
    toggleLight(on: boolean, cfg: SunriseTimerConfig): void
}
