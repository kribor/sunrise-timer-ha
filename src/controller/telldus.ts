import {getGot} from "@naturalcycles/nodejs-lib";
import {Controller} from "./controller";
import {SunriseTimerConfig} from "../svc/config.service";
import OAuth = require("oauth-1.0a");
import * as crypto from "crypto";
import {Token} from "oauth-1.0a";

const oauth = new OAuth({
  consumer: {
    key: process.env['TELLDUS_CONSUMER_KEY'] as string,
    secret: process.env['TELLDUS_CONSUMER_SECRET'] as string
  },
  signature_method: 'HMAC-SHA1',
  hash_function: (baseString, key) => crypto.createHmac('sha1', key).update(baseString).digest('base64')
});

const token = {
  key: process.env['TELLDUS_ACCESS_TOKEN'],
  secret: process.env['TELLDUS_ACCESS_TOKEN_SECRET']
} as Token;

const baseUrl = 'https://pa-api.telldus.com/json'

const telldusGot = getGot({
    logStart: true,
    timeout: 30_000,
})

interface TelldusStatus {
    status: string
}

export class TelldusController implements Controller {
    constructor() {}

    async dimLight(dimLevel: number, cfg: SunriseTimerConfig) : Promise<void> {
        const url = `${baseUrl}/device/dim?id=${cfg.telldus?.dimmer_id}&level=${Math.floor(dimLevel * 255 / 100)}`

        const r = await telldusGot.get(
          url,
          { headers: oauth.toHeader(oauth.authorize({url, method: 'GET'}, token)) as any,}
        ).json<TelldusStatus>()
        if (r.status !== 'success') {
            throw Error(`dimLight received unexpected response: ${r}`)
        }
    }

    async toggleLight(on: boolean, cfg: SunriseTimerConfig) : Promise<void> {
        const url: string = on ? `${baseUrl}/device/turnOn?id=${cfg.telldus?.dimmer_id}` : `${baseUrl}/device/turnOff?id=${cfg.telldus?.dimmer_id}`

        const r = await telldusGot.get(
          url,
          { headers: oauth.toHeader(oauth.authorize({url, method: 'GET'}, token)) as any }
        ).json<TelldusStatus>()
        if (r.status !== 'success') {
            throw Error(`dimLight received unexpected response: ${r}`)
        }
    }
}

export const controller = new TelldusController()
