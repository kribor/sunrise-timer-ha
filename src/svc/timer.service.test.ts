import {dayjs, IDayjs} from "@naturalcycles/time-lib";
import {_singleExec, determineDimLevel, run} from "./timer.service";
import {config, SunriseTimerConfig} from "./config.service";
import {mockTime} from "@naturalcycles/dev-lib/dist/testing";
import {controller} from "../controller/telldus";
import * as yaml from "js-yaml";
import * as fs from 'fs'
import {pMap} from "@naturalcycles/js-lib";

beforeEach(() => {
  jest.restoreAllMocks()
  mockTime(dayjs('2021-11-28').unix())
})

describe('timer.service with configs', () => {
  const testCases: [string, SunriseTimerConfig][] = [
    // Req: Track to fb with db data if user is logged in and paid less than 7 days ago
    [
      'Default config',
      yaml.load(fs.readFileSync('config-defaults.yaml', 'utf8')) as SunriseTimerConfig,
    ],[
      'Custom dimmer interval config',
      {
        ...(yaml.load(fs.readFileSync('config-defaults.yaml', 'utf8')) as SunriseTimerConfig),
        dimmer_min: 10,
        dimmer_max: 90
      }
    ],
  ]
  test.each(testCases)('%s', async (_, cfg) => {

    jest.spyOn(controller, 'toggleLight').mockImplementation()
    jest.spyOn(controller, 'dimLight').mockImplementation()

    //Array with every 5 minutes of day
    const totalMinutes = [...Array(60*24/5)].map((_, i) => i * 5);

    const dayJsArr: IDayjs[] = totalMinutes.map(totMin => dayjs().hour(Math.floor(totMin / 60)).minute(totMin % 60))

    const dimLevelArr = await pMap(dayJsArr, async(now) => {
      mockTime(now.unix());
      return [now.format("HH:mm"), await _singleExec(controller, cfg)]
    })

    expect(dimLevelArr).toMatchSnapshot()

  })
})
