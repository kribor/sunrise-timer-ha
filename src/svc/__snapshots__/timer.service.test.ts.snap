// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`timer.service with configs Custom dimmer interval config 1`] = `
Array [
  Array [
    "00:00",
    0,
  ],
  Array [
    "00:05",
    0,
  ],
  Array [
    "00:10",
    0,
  ],
  Array [
    "00:15",
    0,
  ],
  Array [
    "00:20",
    0,
  ],
  Array [
    "00:25",
    0,
  ],
  Array [
    "00:30",
    0,
  ],
  Array [
    "00:35",
    0,
  ],
  Array [
    "00:40",
    0,
  ],
  Array [
    "00:45",
    0,
  ],
  Array [
    "00:50",
    0,
  ],
  Array [
    "00:55",
    0,
  ],
  Array [
    "01:00",
    0,
  ],
  Array [
    "01:05",
    0,
  ],
  Array [
    "01:10",
    0,
  ],
  Array [
    "01:15",
    0,
  ],
  Array [
    "01:20",
    0,
  ],
  Array [
    "01:25",
    0,
  ],
  Array [
    "01:30",
    0,
  ],
  Array [
    "01:35",
    0,
  ],
  Array [
    "01:40",
    0,
  ],
  Array [
    "01:45",
    0,
  ],
  Array [
    "01:50",
    0,
  ],
  Array [
    "01:55",
    0,
  ],
  Array [
    "02:00",
    0,
  ],
  Array [
    "02:05",
    0,
  ],
  Array [
    "02:10",
    0,
  ],
  Array [
    "02:15",
    0,
  ],
  Array [
    "02:20",
    0,
  ],
  Array [
    "02:25",
    0,
  ],
  Array [
    "02:30",
    0,
  ],
  Array [
    "02:35",
    0,
  ],
  Array [
    "02:40",
    0,
  ],
  Array [
    "02:45",
    0,
  ],
  Array [
    "02:50",
    0,
  ],
  Array [
    "02:55",
    0,
  ],
  Array [
    "03:00",
    0,
  ],
  Array [
    "03:05",
    0,
  ],
  Array [
    "03:10",
    0,
  ],
  Array [
    "03:15",
    0,
  ],
  Array [
    "03:20",
    0,
  ],
  Array [
    "03:25",
    0,
  ],
  Array [
    "03:30",
    0,
  ],
  Array [
    "03:35",
    0,
  ],
  Array [
    "03:40",
    0,
  ],
  Array [
    "03:45",
    0,
  ],
  Array [
    "03:50",
    0,
  ],
  Array [
    "03:55",
    0,
  ],
  Array [
    "04:00",
    0,
  ],
  Array [
    "04:05",
    12,
  ],
  Array [
    "04:10",
    16,
  ],
  Array [
    "04:15",
    18,
  ],
  Array [
    "04:20",
    22,
  ],
  Array [
    "04:25",
    26,
  ],
  Array [
    "04:30",
    30,
  ],
  Array [
    "04:35",
    32,
  ],
  Array [
    "04:40",
    36,
  ],
  Array [
    "04:45",
    38,
  ],
  Array [
    "04:50",
    42,
  ],
  Array [
    "04:55",
    46,
  ],
  Array [
    "05:00",
    50,
  ],
  Array [
    "05:05",
    52,
  ],
  Array [
    "05:10",
    56,
  ],
  Array [
    "05:15",
    58,
  ],
  Array [
    "05:20",
    62,
  ],
  Array [
    "05:25",
    66,
  ],
  Array [
    "05:30",
    70,
  ],
  Array [
    "05:35",
    72,
  ],
  Array [
    "05:40",
    76,
  ],
  Array [
    "05:45",
    78,
  ],
  Array [
    "05:50",
    82,
  ],
  Array [
    "05:55",
    86,
  ],
  Array [
    "06:00",
    100,
  ],
  Array [
    "06:05",
    100,
  ],
  Array [
    "06:10",
    100,
  ],
  Array [
    "06:15",
    100,
  ],
  Array [
    "06:20",
    100,
  ],
  Array [
    "06:25",
    100,
  ],
  Array [
    "06:30",
    100,
  ],
  Array [
    "06:35",
    100,
  ],
  Array [
    "06:40",
    100,
  ],
  Array [
    "06:45",
    100,
  ],
  Array [
    "06:50",
    100,
  ],
  Array [
    "06:55",
    100,
  ],
  Array [
    "07:00",
    100,
  ],
  Array [
    "07:05",
    100,
  ],
  Array [
    "07:10",
    100,
  ],
  Array [
    "07:15",
    100,
  ],
  Array [
    "07:20",
    100,
  ],
  Array [
    "07:25",
    100,
  ],
  Array [
    "07:30",
    100,
  ],
  Array [
    "07:35",
    100,
  ],
  Array [
    "07:40",
    100,
  ],
  Array [
    "07:45",
    100,
  ],
  Array [
    "07:50",
    100,
  ],
  Array [
    "07:55",
    100,
  ],
  Array [
    "08:00",
    100,
  ],
  Array [
    "08:05",
    100,
  ],
  Array [
    "08:10",
    100,
  ],
  Array [
    "08:15",
    100,
  ],
  Array [
    "08:20",
    100,
  ],
  Array [
    "08:25",
    100,
  ],
  Array [
    "08:30",
    100,
  ],
  Array [
    "08:35",
    100,
  ],
  Array [
    "08:40",
    100,
  ],
  Array [
    "08:45",
    100,
  ],
  Array [
    "08:50",
    100,
  ],
  Array [
    "08:55",
    100,
  ],
  Array [
    "09:00",
    100,
  ],
  Array [
    "09:05",
    100,
  ],
  Array [
    "09:10",
    100,
  ],
  Array [
    "09:15",
    100,
  ],
  Array [
    "09:20",
    0,
  ],
  Array [
    "09:25",
    0,
  ],
  Array [
    "09:30",
    0,
  ],
  Array [
    "09:35",
    0,
  ],
  Array [
    "09:40",
    0,
  ],
  Array [
    "09:45",
    0,
  ],
  Array [
    "09:50",
    0,
  ],
  Array [
    "09:55",
    0,
  ],
  Array [
    "10:00",
    0,
  ],
  Array [
    "10:05",
    0,
  ],
  Array [
    "10:10",
    0,
  ],
  Array [
    "10:15",
    0,
  ],
  Array [
    "10:20",
    0,
  ],
  Array [
    "10:25",
    0,
  ],
  Array [
    "10:30",
    0,
  ],
  Array [
    "10:35",
    0,
  ],
  Array [
    "10:40",
    0,
  ],
  Array [
    "10:45",
    0,
  ],
  Array [
    "10:50",
    0,
  ],
  Array [
    "10:55",
    0,
  ],
  Array [
    "11:00",
    0,
  ],
  Array [
    "11:05",
    0,
  ],
  Array [
    "11:10",
    0,
  ],
  Array [
    "11:15",
    0,
  ],
  Array [
    "11:20",
    0,
  ],
  Array [
    "11:25",
    0,
  ],
  Array [
    "11:30",
    0,
  ],
  Array [
    "11:35",
    0,
  ],
  Array [
    "11:40",
    0,
  ],
  Array [
    "11:45",
    0,
  ],
  Array [
    "11:50",
    0,
  ],
  Array [
    "11:55",
    0,
  ],
  Array [
    "12:00",
    100,
  ],
  Array [
    "12:05",
    100,
  ],
  Array [
    "12:10",
    100,
  ],
  Array [
    "12:15",
    100,
  ],
  Array [
    "12:20",
    100,
  ],
  Array [
    "12:25",
    100,
  ],
  Array [
    "12:30",
    100,
  ],
  Array [
    "12:35",
    100,
  ],
  Array [
    "12:40",
    100,
  ],
  Array [
    "12:45",
    100,
  ],
  Array [
    "12:50",
    100,
  ],
  Array [
    "12:55",
    100,
  ],
  Array [
    "13:00",
    100,
  ],
  Array [
    "13:05",
    100,
  ],
  Array [
    "13:10",
    100,
  ],
  Array [
    "13:15",
    100,
  ],
  Array [
    "13:20",
    100,
  ],
  Array [
    "13:25",
    100,
  ],
  Array [
    "13:30",
    100,
  ],
  Array [
    "13:35",
    100,
  ],
  Array [
    "13:40",
    100,
  ],
  Array [
    "13:45",
    100,
  ],
  Array [
    "13:50",
    100,
  ],
  Array [
    "13:55",
    100,
  ],
  Array [
    "14:00",
    100,
  ],
  Array [
    "14:05",
    100,
  ],
  Array [
    "14:10",
    100,
  ],
  Array [
    "14:15",
    100,
  ],
  Array [
    "14:20",
    100,
  ],
  Array [
    "14:25",
    100,
  ],
  Array [
    "14:30",
    100,
  ],
  Array [
    "14:35",
    100,
  ],
  Array [
    "14:40",
    100,
  ],
  Array [
    "14:45",
    100,
  ],
  Array [
    "14:50",
    100,
  ],
  Array [
    "14:55",
    100,
  ],
  Array [
    "15:00",
    100,
  ],
  Array [
    "15:05",
    100,
  ],
  Array [
    "15:10",
    100,
  ],
  Array [
    "15:15",
    100,
  ],
  Array [
    "15:20",
    100,
  ],
  Array [
    "15:25",
    100,
  ],
  Array [
    "15:30",
    100,
  ],
  Array [
    "15:35",
    100,
  ],
  Array [
    "15:40",
    100,
  ],
  Array [
    "15:45",
    100,
  ],
  Array [
    "15:50",
    100,
  ],
  Array [
    "15:55",
    100,
  ],
  Array [
    "16:00",
    0,
  ],
  Array [
    "16:05",
    86,
  ],
  Array [
    "16:10",
    82,
  ],
  Array [
    "16:15",
    78,
  ],
  Array [
    "16:20",
    76,
  ],
  Array [
    "16:25",
    72,
  ],
  Array [
    "16:30",
    70,
  ],
  Array [
    "16:35",
    66,
  ],
  Array [
    "16:40",
    62,
  ],
  Array [
    "16:45",
    58,
  ],
  Array [
    "16:50",
    56,
  ],
  Array [
    "16:55",
    52,
  ],
  Array [
    "17:00",
    50,
  ],
  Array [
    "17:05",
    46,
  ],
  Array [
    "17:10",
    42,
  ],
  Array [
    "17:15",
    38,
  ],
  Array [
    "17:20",
    36,
  ],
  Array [
    "17:25",
    32,
  ],
  Array [
    "17:30",
    30,
  ],
  Array [
    "17:35",
    26,
  ],
  Array [
    "17:40",
    22,
  ],
  Array [
    "17:45",
    18,
  ],
  Array [
    "17:50",
    16,
  ],
  Array [
    "17:55",
    12,
  ],
  Array [
    "18:00",
    0,
  ],
  Array [
    "18:05",
    0,
  ],
  Array [
    "18:10",
    0,
  ],
  Array [
    "18:15",
    0,
  ],
  Array [
    "18:20",
    0,
  ],
  Array [
    "18:25",
    0,
  ],
  Array [
    "18:30",
    0,
  ],
  Array [
    "18:35",
    0,
  ],
  Array [
    "18:40",
    0,
  ],
  Array [
    "18:45",
    0,
  ],
  Array [
    "18:50",
    0,
  ],
  Array [
    "18:55",
    0,
  ],
  Array [
    "19:00",
    0,
  ],
  Array [
    "19:05",
    0,
  ],
  Array [
    "19:10",
    0,
  ],
  Array [
    "19:15",
    0,
  ],
  Array [
    "19:20",
    0,
  ],
  Array [
    "19:25",
    0,
  ],
  Array [
    "19:30",
    0,
  ],
  Array [
    "19:35",
    0,
  ],
  Array [
    "19:40",
    0,
  ],
  Array [
    "19:45",
    0,
  ],
  Array [
    "19:50",
    0,
  ],
  Array [
    "19:55",
    0,
  ],
  Array [
    "20:00",
    0,
  ],
  Array [
    "20:05",
    0,
  ],
  Array [
    "20:10",
    0,
  ],
  Array [
    "20:15",
    0,
  ],
  Array [
    "20:20",
    0,
  ],
  Array [
    "20:25",
    0,
  ],
  Array [
    "20:30",
    0,
  ],
  Array [
    "20:35",
    0,
  ],
  Array [
    "20:40",
    0,
  ],
  Array [
    "20:45",
    0,
  ],
  Array [
    "20:50",
    0,
  ],
  Array [
    "20:55",
    0,
  ],
  Array [
    "21:00",
    0,
  ],
  Array [
    "21:05",
    0,
  ],
  Array [
    "21:10",
    0,
  ],
  Array [
    "21:15",
    0,
  ],
  Array [
    "21:20",
    0,
  ],
  Array [
    "21:25",
    0,
  ],
  Array [
    "21:30",
    0,
  ],
  Array [
    "21:35",
    0,
  ],
  Array [
    "21:40",
    0,
  ],
  Array [
    "21:45",
    0,
  ],
  Array [
    "21:50",
    0,
  ],
  Array [
    "21:55",
    0,
  ],
  Array [
    "22:00",
    0,
  ],
  Array [
    "22:05",
    0,
  ],
  Array [
    "22:10",
    0,
  ],
  Array [
    "22:15",
    0,
  ],
  Array [
    "22:20",
    0,
  ],
  Array [
    "22:25",
    0,
  ],
  Array [
    "22:30",
    0,
  ],
  Array [
    "22:35",
    0,
  ],
  Array [
    "22:40",
    0,
  ],
  Array [
    "22:45",
    0,
  ],
  Array [
    "22:50",
    0,
  ],
  Array [
    "22:55",
    0,
  ],
  Array [
    "23:00",
    0,
  ],
  Array [
    "23:05",
    0,
  ],
  Array [
    "23:10",
    0,
  ],
  Array [
    "23:15",
    0,
  ],
  Array [
    "23:20",
    0,
  ],
  Array [
    "23:25",
    0,
  ],
  Array [
    "23:30",
    0,
  ],
  Array [
    "23:35",
    0,
  ],
  Array [
    "23:40",
    0,
  ],
  Array [
    "23:45",
    0,
  ],
  Array [
    "23:50",
    0,
  ],
  Array [
    "23:55",
    0,
  ],
]
`;

exports[`timer.service with configs Default config 1`] = `
Array [
  Array [
    "00:00",
    0,
  ],
  Array [
    "00:05",
    0,
  ],
  Array [
    "00:10",
    0,
  ],
  Array [
    "00:15",
    0,
  ],
  Array [
    "00:20",
    0,
  ],
  Array [
    "00:25",
    0,
  ],
  Array [
    "00:30",
    0,
  ],
  Array [
    "00:35",
    0,
  ],
  Array [
    "00:40",
    0,
  ],
  Array [
    "00:45",
    0,
  ],
  Array [
    "00:50",
    0,
  ],
  Array [
    "00:55",
    0,
  ],
  Array [
    "01:00",
    0,
  ],
  Array [
    "01:05",
    0,
  ],
  Array [
    "01:10",
    0,
  ],
  Array [
    "01:15",
    0,
  ],
  Array [
    "01:20",
    0,
  ],
  Array [
    "01:25",
    0,
  ],
  Array [
    "01:30",
    0,
  ],
  Array [
    "01:35",
    0,
  ],
  Array [
    "01:40",
    0,
  ],
  Array [
    "01:45",
    0,
  ],
  Array [
    "01:50",
    0,
  ],
  Array [
    "01:55",
    0,
  ],
  Array [
    "02:00",
    0,
  ],
  Array [
    "02:05",
    0,
  ],
  Array [
    "02:10",
    0,
  ],
  Array [
    "02:15",
    0,
  ],
  Array [
    "02:20",
    0,
  ],
  Array [
    "02:25",
    0,
  ],
  Array [
    "02:30",
    0,
  ],
  Array [
    "02:35",
    0,
  ],
  Array [
    "02:40",
    0,
  ],
  Array [
    "02:45",
    0,
  ],
  Array [
    "02:50",
    0,
  ],
  Array [
    "02:55",
    0,
  ],
  Array [
    "03:00",
    0,
  ],
  Array [
    "03:05",
    0,
  ],
  Array [
    "03:10",
    0,
  ],
  Array [
    "03:15",
    0,
  ],
  Array [
    "03:20",
    0,
  ],
  Array [
    "03:25",
    0,
  ],
  Array [
    "03:30",
    0,
  ],
  Array [
    "03:35",
    0,
  ],
  Array [
    "03:40",
    0,
  ],
  Array [
    "03:45",
    0,
  ],
  Array [
    "03:50",
    0,
  ],
  Array [
    "03:55",
    0,
  ],
  Array [
    "04:00",
    0,
  ],
  Array [
    "04:05",
    4,
  ],
  Array [
    "04:10",
    8,
  ],
  Array [
    "04:15",
    12,
  ],
  Array [
    "04:20",
    16,
  ],
  Array [
    "04:25",
    20,
  ],
  Array [
    "04:30",
    24,
  ],
  Array [
    "04:35",
    28,
  ],
  Array [
    "04:40",
    32,
  ],
  Array [
    "04:45",
    36,
  ],
  Array [
    "04:50",
    40,
  ],
  Array [
    "04:55",
    44,
  ],
  Array [
    "05:00",
    50,
  ],
  Array [
    "05:05",
    54,
  ],
  Array [
    "05:10",
    58,
  ],
  Array [
    "05:15",
    62,
  ],
  Array [
    "05:20",
    66,
  ],
  Array [
    "05:25",
    70,
  ],
  Array [
    "05:30",
    74,
  ],
  Array [
    "05:35",
    78,
  ],
  Array [
    "05:40",
    82,
  ],
  Array [
    "05:45",
    86,
  ],
  Array [
    "05:50",
    90,
  ],
  Array [
    "05:55",
    94,
  ],
  Array [
    "06:00",
    100,
  ],
  Array [
    "06:05",
    100,
  ],
  Array [
    "06:10",
    100,
  ],
  Array [
    "06:15",
    100,
  ],
  Array [
    "06:20",
    100,
  ],
  Array [
    "06:25",
    100,
  ],
  Array [
    "06:30",
    100,
  ],
  Array [
    "06:35",
    100,
  ],
  Array [
    "06:40",
    100,
  ],
  Array [
    "06:45",
    100,
  ],
  Array [
    "06:50",
    100,
  ],
  Array [
    "06:55",
    100,
  ],
  Array [
    "07:00",
    100,
  ],
  Array [
    "07:05",
    100,
  ],
  Array [
    "07:10",
    100,
  ],
  Array [
    "07:15",
    100,
  ],
  Array [
    "07:20",
    100,
  ],
  Array [
    "07:25",
    100,
  ],
  Array [
    "07:30",
    100,
  ],
  Array [
    "07:35",
    100,
  ],
  Array [
    "07:40",
    100,
  ],
  Array [
    "07:45",
    100,
  ],
  Array [
    "07:50",
    100,
  ],
  Array [
    "07:55",
    100,
  ],
  Array [
    "08:00",
    100,
  ],
  Array [
    "08:05",
    100,
  ],
  Array [
    "08:10",
    100,
  ],
  Array [
    "08:15",
    100,
  ],
  Array [
    "08:20",
    100,
  ],
  Array [
    "08:25",
    100,
  ],
  Array [
    "08:30",
    100,
  ],
  Array [
    "08:35",
    100,
  ],
  Array [
    "08:40",
    100,
  ],
  Array [
    "08:45",
    100,
  ],
  Array [
    "08:50",
    100,
  ],
  Array [
    "08:55",
    100,
  ],
  Array [
    "09:00",
    100,
  ],
  Array [
    "09:05",
    100,
  ],
  Array [
    "09:10",
    100,
  ],
  Array [
    "09:15",
    100,
  ],
  Array [
    "09:20",
    0,
  ],
  Array [
    "09:25",
    0,
  ],
  Array [
    "09:30",
    0,
  ],
  Array [
    "09:35",
    0,
  ],
  Array [
    "09:40",
    0,
  ],
  Array [
    "09:45",
    0,
  ],
  Array [
    "09:50",
    0,
  ],
  Array [
    "09:55",
    0,
  ],
  Array [
    "10:00",
    0,
  ],
  Array [
    "10:05",
    0,
  ],
  Array [
    "10:10",
    0,
  ],
  Array [
    "10:15",
    0,
  ],
  Array [
    "10:20",
    0,
  ],
  Array [
    "10:25",
    0,
  ],
  Array [
    "10:30",
    0,
  ],
  Array [
    "10:35",
    0,
  ],
  Array [
    "10:40",
    0,
  ],
  Array [
    "10:45",
    0,
  ],
  Array [
    "10:50",
    0,
  ],
  Array [
    "10:55",
    0,
  ],
  Array [
    "11:00",
    0,
  ],
  Array [
    "11:05",
    0,
  ],
  Array [
    "11:10",
    0,
  ],
  Array [
    "11:15",
    0,
  ],
  Array [
    "11:20",
    0,
  ],
  Array [
    "11:25",
    0,
  ],
  Array [
    "11:30",
    0,
  ],
  Array [
    "11:35",
    0,
  ],
  Array [
    "11:40",
    0,
  ],
  Array [
    "11:45",
    0,
  ],
  Array [
    "11:50",
    0,
  ],
  Array [
    "11:55",
    0,
  ],
  Array [
    "12:00",
    100,
  ],
  Array [
    "12:05",
    100,
  ],
  Array [
    "12:10",
    100,
  ],
  Array [
    "12:15",
    100,
  ],
  Array [
    "12:20",
    100,
  ],
  Array [
    "12:25",
    100,
  ],
  Array [
    "12:30",
    100,
  ],
  Array [
    "12:35",
    100,
  ],
  Array [
    "12:40",
    100,
  ],
  Array [
    "12:45",
    100,
  ],
  Array [
    "12:50",
    100,
  ],
  Array [
    "12:55",
    100,
  ],
  Array [
    "13:00",
    100,
  ],
  Array [
    "13:05",
    100,
  ],
  Array [
    "13:10",
    100,
  ],
  Array [
    "13:15",
    100,
  ],
  Array [
    "13:20",
    100,
  ],
  Array [
    "13:25",
    100,
  ],
  Array [
    "13:30",
    100,
  ],
  Array [
    "13:35",
    100,
  ],
  Array [
    "13:40",
    100,
  ],
  Array [
    "13:45",
    100,
  ],
  Array [
    "13:50",
    100,
  ],
  Array [
    "13:55",
    100,
  ],
  Array [
    "14:00",
    100,
  ],
  Array [
    "14:05",
    100,
  ],
  Array [
    "14:10",
    100,
  ],
  Array [
    "14:15",
    100,
  ],
  Array [
    "14:20",
    100,
  ],
  Array [
    "14:25",
    100,
  ],
  Array [
    "14:30",
    100,
  ],
  Array [
    "14:35",
    100,
  ],
  Array [
    "14:40",
    100,
  ],
  Array [
    "14:45",
    100,
  ],
  Array [
    "14:50",
    100,
  ],
  Array [
    "14:55",
    100,
  ],
  Array [
    "15:00",
    100,
  ],
  Array [
    "15:05",
    100,
  ],
  Array [
    "15:10",
    100,
  ],
  Array [
    "15:15",
    100,
  ],
  Array [
    "15:20",
    100,
  ],
  Array [
    "15:25",
    100,
  ],
  Array [
    "15:30",
    100,
  ],
  Array [
    "15:35",
    100,
  ],
  Array [
    "15:40",
    100,
  ],
  Array [
    "15:45",
    100,
  ],
  Array [
    "15:50",
    100,
  ],
  Array [
    "15:55",
    100,
  ],
  Array [
    "16:00",
    0,
  ],
  Array [
    "16:05",
    94,
  ],
  Array [
    "16:10",
    90,
  ],
  Array [
    "16:15",
    86,
  ],
  Array [
    "16:20",
    82,
  ],
  Array [
    "16:25",
    78,
  ],
  Array [
    "16:30",
    74,
  ],
  Array [
    "16:35",
    70,
  ],
  Array [
    "16:40",
    66,
  ],
  Array [
    "16:45",
    62,
  ],
  Array [
    "16:50",
    58,
  ],
  Array [
    "16:55",
    54,
  ],
  Array [
    "17:00",
    50,
  ],
  Array [
    "17:05",
    44,
  ],
  Array [
    "17:10",
    40,
  ],
  Array [
    "17:15",
    36,
  ],
  Array [
    "17:20",
    32,
  ],
  Array [
    "17:25",
    28,
  ],
  Array [
    "17:30",
    24,
  ],
  Array [
    "17:35",
    20,
  ],
  Array [
    "17:40",
    16,
  ],
  Array [
    "17:45",
    12,
  ],
  Array [
    "17:50",
    8,
  ],
  Array [
    "17:55",
    4,
  ],
  Array [
    "18:00",
    0,
  ],
  Array [
    "18:05",
    0,
  ],
  Array [
    "18:10",
    0,
  ],
  Array [
    "18:15",
    0,
  ],
  Array [
    "18:20",
    0,
  ],
  Array [
    "18:25",
    0,
  ],
  Array [
    "18:30",
    0,
  ],
  Array [
    "18:35",
    0,
  ],
  Array [
    "18:40",
    0,
  ],
  Array [
    "18:45",
    0,
  ],
  Array [
    "18:50",
    0,
  ],
  Array [
    "18:55",
    0,
  ],
  Array [
    "19:00",
    0,
  ],
  Array [
    "19:05",
    0,
  ],
  Array [
    "19:10",
    0,
  ],
  Array [
    "19:15",
    0,
  ],
  Array [
    "19:20",
    0,
  ],
  Array [
    "19:25",
    0,
  ],
  Array [
    "19:30",
    0,
  ],
  Array [
    "19:35",
    0,
  ],
  Array [
    "19:40",
    0,
  ],
  Array [
    "19:45",
    0,
  ],
  Array [
    "19:50",
    0,
  ],
  Array [
    "19:55",
    0,
  ],
  Array [
    "20:00",
    0,
  ],
  Array [
    "20:05",
    0,
  ],
  Array [
    "20:10",
    0,
  ],
  Array [
    "20:15",
    0,
  ],
  Array [
    "20:20",
    0,
  ],
  Array [
    "20:25",
    0,
  ],
  Array [
    "20:30",
    0,
  ],
  Array [
    "20:35",
    0,
  ],
  Array [
    "20:40",
    0,
  ],
  Array [
    "20:45",
    0,
  ],
  Array [
    "20:50",
    0,
  ],
  Array [
    "20:55",
    0,
  ],
  Array [
    "21:00",
    0,
  ],
  Array [
    "21:05",
    0,
  ],
  Array [
    "21:10",
    0,
  ],
  Array [
    "21:15",
    0,
  ],
  Array [
    "21:20",
    0,
  ],
  Array [
    "21:25",
    0,
  ],
  Array [
    "21:30",
    0,
  ],
  Array [
    "21:35",
    0,
  ],
  Array [
    "21:40",
    0,
  ],
  Array [
    "21:45",
    0,
  ],
  Array [
    "21:50",
    0,
  ],
  Array [
    "21:55",
    0,
  ],
  Array [
    "22:00",
    0,
  ],
  Array [
    "22:05",
    0,
  ],
  Array [
    "22:10",
    0,
  ],
  Array [
    "22:15",
    0,
  ],
  Array [
    "22:20",
    0,
  ],
  Array [
    "22:25",
    0,
  ],
  Array [
    "22:30",
    0,
  ],
  Array [
    "22:35",
    0,
  ],
  Array [
    "22:40",
    0,
  ],
  Array [
    "22:45",
    0,
  ],
  Array [
    "22:50",
    0,
  ],
  Array [
    "22:55",
    0,
  ],
  Array [
    "23:00",
    0,
  ],
  Array [
    "23:05",
    0,
  ],
  Array [
    "23:10",
    0,
  ],
  Array [
    "23:15",
    0,
  ],
  Array [
    "23:20",
    0,
  ],
  Array [
    "23:25",
    0,
  ],
  Array [
    "23:30",
    0,
  ],
  Array [
    "23:35",
    0,
  ],
  Array [
    "23:40",
    0,
  ],
  Array [
    "23:45",
    0,
  ],
  Array [
    "23:50",
    0,
  ],
  Array [
    "23:55",
    0,
  ],
]
`;
